govc/task/recent.go

```diff
@@ -101,7 +124,112 @@ func taskName(info *types.TaskInfo) string {
 	}
 }
 
+type history struct {
+	*task.HistoryCollector
+
+	cmd *recent
+}
+
+func (h *history) Collect(ctx context.Context, f func([]types.TaskInfo)) error {
+	for {
+		tasks, err := h.ReadNextTasks(ctx, 10)
+		if err != nil {
+			return err
+		}
+
+		if len(tasks) == 0 {
+			if h.cmd.follow {
+				// TODO: this only follows new events.
+				// need to watch TaskHistoryCollector.LatestPage for updates to existing Tasks
+				time.Sleep(time.Second)
+				continue
+			}
+			break
+		}
+
+		f(tasks)
+	}
+	return nil
+}
+
+type collector interface {
+	Collect(context.Context, func([]types.TaskInfo)) error
+	Destroy(context.Context) error
+}
+
+// useRecent returns true if any options are specified that require use of TaskHistoryCollector
+func (cmd *recent) useRecent() bool {
+	return cmd.begin == 0 && cmd.end == 0 && !cmd.r && len(cmd.state) == 0
+}
+
+func (cmd *recent) newCollector(ctx context.Context, c *vim25.Client, ref *types.ManagedObjectReference) (collector, error) {
+	if cmd.useRecent() {
+		// original flavor of this command that uses `RecentTask` instead of `TaskHistoryCollector`
+		if ref == nil {
+			ref = c.ServiceContent.TaskManager
+		}
+
+		v, err := view.NewManager(c).CreateTaskView(ctx, ref)
+		if err != nil {
+			return nil, err
+		}
+
+		v.Follow = cmd.follow && cmd.plain
+		return v, nil
+	}
+
+	m := task.NewManager(c)
+	r := types.TaskFilterSpecRecursionOptionSelf
+	if ref == nil {
+		ref = &c.ServiceContent.RootFolder
+		cmd.r = true
+	}
+
+	now, err := methods.GetCurrentTime(ctx, c) // vCenter server time (UTC)
+	if err != nil {
+		return nil, err
+	}
+
+	if cmd.r {
+		r = types.TaskFilterSpecRecursionOptionAll
+	}
+
+	if cmd.begin == 0 {
+		cmd.begin = 10 * time.Minute
+	}
+
+	filter := types.TaskFilterSpec{
+		Entity: &types.TaskFilterSpecByEntity{
+			Entity:    *ref,
+			Recursion: r,
+		},
+		Time: &types.TaskFilterSpecByTime{
+			TimeType:  types.TaskFilterSpecTimeOptionStartedTime,
+			BeginTime: types.NewTime(now.Add(cmd.begin * -1)),
```

Comment:
What does multiplying this by `-1` do? Same question for the end time below.

Reply:
Example, given `-b 1h`, we want to show tasks from the past hour.
https://pkg.go.dev/time#Time.Sub
> To compute t-d for a duration d, use t.Add(-d)

Tho reading that changed to `-cmd.begin` instead of `* -1`